### ISSUE TITLE
Show statistics for target package is restored for

### ISSUE THREAD

**Issue by twsouthwick** [NuGet/NuGetGallery#3450]
Generally, packages should not have a separate package for a target (such as .NET 4.0) but instead have it as another dependency within the package. However, by keeping it separate, statistics about usage can be easily seen (ie - how many people are using the .NET 4.0 version). Is there a way to add visualization for that to the website (or some other way to get at that data)?

**Comment by xavierdecoster** [comment]
I don't think we can offer a solid story today in response to this question based on the download data we already collect and use to show statistics. The server doesn't know about the target framework selected by the client.

As for showing what package caused the dependent package to be installed: I think older or `NuGet.Core`-based clients do send this information in custom HTTP-headers (e.g. [`NuGet-Operation`](https://github.com/NuGet/NuGet2/blob/e89c4b2fc6dc2fc752d2fceecd2a2175b0f4ab69/src/Core/Http/RepositoryOperationNames.cs#L8), [`NuGet-DependentPackage`](https://github.com/NuGet/NuGet2/blob/e89c4b2fc6dc2fc752d2fceecd2a2175b0f4ab69/src/Core/Http/RepositoryOperationNames.cs#L9), and [`NuGet-DependentPackageVersion`](https://github.com/NuGet/NuGet2/blob/e89c4b2fc6dc2fc752d2fceecd2a2175b0f4ab69/src/Core/Http/RepositoryOperationNames.cs#L10)), but not sure about support in v3 protocol and newer clients. 

Either way, the stats processing pipeline does not collect that information today.

@rrelyea this may be a good candidate for client telemetry?

**Comment by reisenberger** [comment]
Thanks @xavierdecoster for feedback.  @twsouthwick has [also suggested a workround](https://github.com/App-vNext/Polly/pull/195#issuecomment-271342191) for us to maintain both a unified package (for users) and split-out package (for stats) at the same time.


@xavierdecoster Not sure if related, but [our Nuget package stats](https://www.nuget.org/packages/Polly) broke around the time this was investigated: the **Total Downloads** figure went up just one from 450,947 to 450,948, then has stuck at that for 5 days or so.  This is unfeasible: we wouldn't just drop from >300 downloads/day to zero.  Could our stats feed have been broken, by someone checking out this issue?

(Raising here as frozen stats co-timed with this issue - but let me know if you want a separate card)

( /cc @joelhulen for info )

**Comment by xavierdecoster** [comment]
@reisenberger thanks for raising. However, as an FYI, the stats not being updated currently is unrelated to this issue (timing is a coincidence), and is already being investigated at the moment.

**Comment by reisenberger** [comment]
Thanks @xavierdecoster @skofman1 for your input and help.

**Comment by rprouse** [comment]
I manage the NUnit package which is very popular and has many targets. It would be extremely useful to us to know which target frameworks users install our package for. Our code base is overly complicated by the number of targets we support, so full stats would allow us to drop support for .NET frameworks with low usage.

**Comment by holytshirt** [comment]
I'm a library author and not having stats on which targets our packages are been restored to makes it super difficult to work out what to continue supporting we are basically blind.

**Comment by jzabroski** [comment]
I suggested a feature like this to Becca Z as part of GitHub Discussions Beta.
In general we need more statistics on our users.
Even better would be namespace/API coverage when someone does a pack on my library.

**Comment by JonDouglas** [comment]
hey package authors,

as far as i'm seeing in this issue, it is suggested people mostly want statistics per TFM restored. i've also heard people want extended durations of stats beyond the 6 weeks.

what other scenarios would you like to see? here is the detailed statistics page for an example package today:

https://www.nuget.org/stats/packages/Newtonsoft.Json?groupby=Version&groupby=ClientName&groupby=ClientVersion

are there any examples of other ecosystems that provide really good statistics that we can be inspired by? leave your comments here

**Comment by jzabroski** [comment]
* Asset Excludes/Includes _if overridden_ e.g. `<PrivateAssets>all</PrivateAssets><IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>`
* Whether it is a Direct vs. Indirect (Transitive) dependency
    * This tells me if the package popularity is derived from another package's success, and potentially invites me to want to help that project that is helping my project gain steam.  I definitely don't want to break my biggest users, at least not without communicating to them the lift and maybe even offering to help with any lift required by breaking changes.
* TFM
* TFM Constraints - Whether TFM was chosen via AssetTargetFallback or due to transitive constraint e.g., project is netstandard2.0, but the "endpoint executable" is 
    * I'm not sure I really want to tightly couple to AssetTargetFallback, as I could see that concept becoming deprecated over time but some way to understand if, we support net48 and netstandard2.0, what version people actually use to target net48  
* SDK type - this would tell us if its being used in a Web App, etc https://learn.microsoft.com/en-us/dotnet/core/project-sdk/overview#available-sdks - it will not tell us if the library is being used in Azure Functions, at least not today as Azure Functions does not require a special SDK type - ultimately what I really want to know is what is the endpoint people are using to package their application with. This is extremely valuable as I often get support requests for things I know nothing about like Azure Functions, and I don't actually know how many users care about support for this. As an open source author, I don't want to spend time on things that one user is using just because they can't get my library to work in some special deployment like Azure Functions. TFM with notes like Unity would be absurdly helpful, but is only half the battle in understanding where people deploy the code to.
* Whether it's used in a .NET CLI Tool or not
    * this is the lowest value of all the suggestions above

**Comment by jzabroski** [comment]
I can also see C# [LangVersion](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/configure-language-version) being useful although very tightly coupled to C#. I'm not sure how F# developers would feel about this but they primarily use paket as a front-end. The usefulness would be in targeting APIs that lend themselves well to certain idioms.

**Comment by jzabroski** [comment]
Also, I have to say, this is my favorite view so far: https://www.nuget.org/stats/packages/FluentMigrator?groupby=Version

![image](https://github.com/user-attachments/assets/9b3259c6-f909-47d0-80ab-cda9f10e4d60)

If this is telling me what I think it is, it's saying I have no downloads in the last six weeks for anything older than 3.2.7 - that alone is really useful information, so thank you!

**Comment by holytshirt** [comment]
TFM please!
The sdk type would be interesting too

**Comment by SimonCropp** [comment]
perhaps the raw data could published somewhere. and then people could leverage how they want

**Comment by tonyqus** [comment]
It's already Apr, 2025.  Any progress on this?
